x-pack/plugins/fleet/public/applications/fleet/sections/agents/agent_details_page/components/agent_details/input_status_utils.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { i18n } from '@kbn/i18n';
+
+import type { PackagePolicy } from '../../../../../types';
+import type {
+  FleetServerAgentComponent,
+  FleetServerAgentComponentStatus,
+  FleetServerAgentComponentUnit,
+} from '../../../../../../../../common/types/models/agent';
+export class InputStatusFormatter {
+  public status?: FleetServerAgentComponentStatus;
+  public description?: string;
+  public hasError?: boolean;
+
+  constructor(status?: FleetServerAgentComponentStatus, message?: string) {
+    this.status = status;
+    this.description =
+      message ||
+      i18n.translate('xpack.fleet.agentDetailsIntegrations.inputStatusDefaultDescription', {
+        defaultMessage: 'Not available',
+      });
+    this.hasError = status === 'FAILED' || status === 'DEGRADED';
+  }
+
+  getErrorTitleFromStatus(): string | undefined {
+    switch (this.status) {
+      case 'FAILED':
+        return i18n.translate('xpack.fleet.agentDetailsIntegrations.inputErrorTitle.failed', {
+          defaultMessage: 'Failed',
+        });
+      case 'DEGRADED':
+        return i18n.translate('xpack.fleet.agentDetailsIntegrations.inputErrorTitle.degraded', {
+          defaultMessage: 'Degraded',
+        });
+      default:
+        return;
+    }
+  }
+}
+
+export const getUnitsByPackage = (
+  agentComponents: FleetServerAgentComponent[],
+  packagePolicy: PackagePolicy
+): FleetServerAgentComponentUnit[] => {
+  const re = new RegExp(`(${packagePolicy.id}|${packagePolicy.package?.name})`);
```

Comment:
I think the regex on package name might return invalid results in case we have a package name that includes the name of another. We could use a stricter regex like `^name/`

Reply:
The Elastic Agent itself only understands inputs, not packages. Any information that would allow tying an input back to a package is injected by Fleet. 

Looks like you have `meta.package` and `package_policy_id` to work with. The input ID has to be unique in the agent policy and looks like it is templated from something like `$inputType-$package-$packagePolicyID` so that might be a good hint for what to use if you want a unique value (you can probably find the code that generates this ID faster than I can :) )

```
  - id: system/metrics-system-4f510cb9-2f4e-4b81-8a19-9969abe1c924
    name: system-1
    revision: 1
    type: system/metrics
    use_output: default
    meta:
      package:
        name: system
        version: 1.31.1
    data_stream:
      namespace: default
    package_policy_id: 4f510cb9-2f4e-4b81-8a19-9969abe1c924
```

Reply:
Thanks for taking a look at this @juliaElastic - this is indeed one of the critical pieces of logic.
Interestingly, the yaml @cmacknz pasted above has a [FullAgentPolicy](https://github.com/elastic/kibana/blob/1d22dcce753eb196a1ed09001cb258e9117c1249/x-pack/plugins/fleet/common/types/models/agent_policy.ts#L93) type which we don't have access without an extra API call. While it _would_ be easier (allowing us to simply retrieve units by input id), I'm not sure it's worth it. I have made a change where we only match the unit id against the package policy id; this means we only retrieve inputs, not outputs. Since we only surface the state of inputs, this should be sufficient for the present requirements.
Please let me know your thoughts.

Reply:
I agree to keep it simple and match on package policy id that works on inputs. We can come back to this later when we add support for outputs.